src/pudl/analysis/allocate_net_gen.py

```diff
@@ -645,30 +1135,501 @@ def calc_allocation_fraction(gen_pm_fuel, drop_interim_cols=True):
+            fuel_alloc["fuel_consumed_for_electricity_mmbtu_gf_tbl"]
+        ),
+    )
+    # do the allocating-ing!
+    fuel_alloc = (
+        fuel_alloc.assign(
+            # we could x.fuel_consumed_mmbtu_bf_tbl.fillna here if we wanted to
+            # take the net gen
+            fuel_consumed_mmbtu=lambda x: x.fuel_consumed_mmbtu_gf_tbl * x.frac,
+            fuel_consumed_for_electricity_mmbtu=lambda x: x.fuel_consumed_for_electricity_mmbtu_gf_tbl
+            * x.frac,
+        )
+        .pipe(apply_pudl_dtypes, group="eia")
+        .dropna(how="all")
+    )
+
+    return fuel_alloc
+
+
+def remove_aggregated_sentinel_value(col: pd.Series, scalar: int = 20) -> pd.Series:
+    """Replace the post-aggregation sentinel values in a column with zero."""
+    return np.where(col.between(0, scalar * MISSING_SENTINEL), 0, col)
+
+
+def group_duplicate_keys(df: pd.DataFrame) -> pd.DataFrame:
+    """Catches duplicate keys in the allocated data and groups them together.
+
+    Merging ``net_gen_alloc`` and ``fuel_alloc`` together requires unique keys in each
+    df. Sometimes the allocation process creates duplicate keys. This function
+    identifies when this happens, and aggregates the data on these keys to remove the
+    duplicates.
+    """
+    # identify any duplicate records
+    duplicate_keys = df[
+        df.duplicated(subset=(IDX_GENS_PM_ESC + ["energy_source_code_num"]))
+    ]
+    # if there are duplicate records, print a warning and fix the issue
+    if len(duplicate_keys) > 0:
+        logger.warning(
+            "Duplicate keys exist in the allocated data."
+            "These will be grouped together, but check the source "
+            "of this issue."
+        )
+        logger.warning(duplicate_keys[IDX_GENS_PM_ESC + ["energy_source_code_num"]])
+
+        df = (
+            df.groupby(IDX_GENS_PM_ESC + ["energy_source_code_num"])
+            .sum(min_count=1)
+            .reset_index()
+        )
+    return df
+
+
+###########################
+# Fuel Allocation Functions
+###########################
+def distribute_annually_reported_data_to_months_if_annual(
+    df: pd.DataFrame,
+    key_columns: list[str],
+    data_column_name: str,
+    freq: Literal["AS", "MS"],
+) -> pd.DataFrame:
+    """Allocates annually-reported data from the gen or bf table to each month.
+
+    Certain plants only report data to the generator table and boiler fuel table
+    on an annual basis. In these cases, their annual total is reported as a single
+    value in January or December, and the other 11 months are reported as missing
+    values. This function first identifies which plants are annual respondents by
+    identifying plants that have 11 months of missing data, with the one month of
+    existing data being in January or December. There are plants that report only one
+    month of data in other months, but over 40% of the one-month data appears in these
```

Comment:
Make the unstated assumption explicit here -- that plants / gens reporting only one month of data in a month that's not January or December actually only had data for that month.

The existence of plants with data only in a single month suggests that we are also inappropriately distributing some truly monthly data that happened to be reported in January and December across all months.

Reply:
it's almost certain we are inappropriately distributing some (albeit probably a very small number) of actually only-one-month reporting data